File: android app/app/src/main/java/com/esibil/call/network/JailsApi.kt
package com.esibil.call.network

import com.esibil.call.Config
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/** Fetches jail/site list from the PPF sites API. */
object JailsApi {

    private val client = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .build()

    suspend fun fetchSites(): List<JailSite> = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(Config.JAILS_API_URL)
            .header("Authorization", "Bearer ${Config.JAILS_API_TOKEN}")
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IllegalStateException("Jails API error: HTTP ${response.code}")
            }
            val body = response.body?.string()
                ?: throw IllegalStateException("Empty jails API response")

            val json = JSONObject(body)
            if (json.optString("status") != "success") {
                throw IllegalStateException("Jails API status: ${json.optString("status")}")
            }

            val sites = json.getJSONArray("sites")
            buildList {
                for (i in 0 until sites.length()) {
                    val item = sites.getJSONObject(i)
                    add(
                        JailSite(
                            ip = item.getString("ip"),
                            jailName = item.getString("jail_name"),
                            status = item.optString("status", "up")
                        )
                    )
                }
            }.sortedBy { it.jailName }
        }
    }
}

File: esibil_call_flutter/lib/services/jails_api.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/app_config.dart';
import '../models/jail_site.dart';

/// Fetches jail/site list from the PPF sites API.
class JailsApi {
  JailsApi({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;

  Future<List<JailSite>> fetchSites() async {
    final response = await _client
        .get(
          Uri.parse(AppConfig.jailsApiUrl),
          headers: {
            'Authorization': 'Bearer ${AppConfig.jailsApiToken}',
          },
        )
        .timeout(const Duration(seconds: 20));

    if (response.statusCode != 200) {
      throw StateError('Jails API error: HTTP ${response.statusCode}');
    }

    final body = response.body;
    if (body.isEmpty) {
      throw StateError('Empty jails API response');
    }

    final json = jsonDecode(body) as Map<String, dynamic>;
    if (json['status'] != 'success') {
      throw StateError('Jails API status: ${json['status']}');
    }

    final sites = json['sites'] as List<dynamic>;
    final list = sites
        .map((e) => JailSite.fromJson(e as Map<String, dynamic>))
        .toList()
      ..sort((a, b) => a.jailName.compareTo(b.jailName));
    return list;
  }
}
